File: algdat_ntnu/sorting/quicksort.cc
#include <iostream>
#include <stdio.h>
#include <random>
#include <chrono>

using std::swap; 
using std::cout; 

int partition(int arr[], int low, int high) {
    int pivot = arr[high]; 
    int from_left = low; 
    int from_right = high - 1; 
    while (true) {
        while (arr[from_left] < pivot) {
            from_left++;
        }; 
        while (from_right > 0 && arr[from_right] > pivot) {
            from_right--; 
        }; 
        if (from_left >= from_right) break; 
        swap(arr[from_left], arr[from_right]); 
    }
    swap(arr[from_left], arr[high]); 
    return from_left; 
}


int quicksort(int arr[], int low, int high) {
    if (high <= low) return high; 
    int pivot = partition(arr, low, high);
    quicksort(arr, low, pivot - 1);
    quicksort(arr, pivot, high); 
    return 0; 
}


int main() {
    int arr[1000000];
    for (int i = 0; i < sizeof(arr)/sizeof(int); i++ ) {
        arr[i] = i; 
    }
    // Randomizing list 
    std::random_device rd; 
    std::mt19937 g(rd()); 
    std::shuffle(std::begin(arr), std::end(arr), g); 

    auto t1 = std::chrono::high_resolution_clock::now(); 
    quicksort(arr, 0, sizeof(arr)/sizeof(int) - 1); 
    auto t2 = std::chrono::high_resolution_clock::now(); 

    auto ms_int = std::chrono::duration_cast<std::chrono::milliseconds>(t2 - t1); 
    cout << ms_int.count() << "ms" << std::endl; 


    return 0; 
}

File: algdat_ntnu/sorting/quick_sort.py
import timeit
from random import sample
from numba import jit

#@jit(nopython=True)
def partition(arr, low, high):
    # Choosing the last element as pivot
    pivot = arr[high]
    from_right = high - 1 # Initial index 
    from_left = low

    while True: 
        while arr[from_left] < pivot: # moving in from the left
            from_left += 1
        while arr[from_right] > pivot and from_right > 0: #moving in from the right
            from_right -= 1
        if from_left >= from_right:
            break
        else: 
            arr[from_left], arr[from_right] = arr[from_right], arr[from_left] # swapping
    arr[from_left], arr[high] = arr[high], arr[from_left] # putting the pivot element in the middle, from the end of the array
    return from_left


#@jit(nopython=True)
def quicksort(arr, low, high): 
    if high <= low:
        return
    pivot = partition(arr, low, high)
    quicksort(arr, low, pivot - 1) 
    quicksort(arr, pivot, high)


integers = [*range(1000000)]
random_integers = sample(integers, len(integers))
integers2 = [*range(1000000)]
random_integers2 = sample(integers, len(integers))

def main():
    quicksort(random_integers2[0:100], 0, len(random_integers[0:100]) - 1)
    print(f"{random_integers[0:100]}")
    time = timeit.timeit('quicksort(random_integers, 0, len(random_integers) - 1)', 'from __main__ import quicksort, random_integers', number=1)
    print(f"time was: {time}")
    


if __name__ == "__main__":
    main()




